contracts/test/GrantRegistry.test.ts

```diff
@@ -0,0 +1,222 @@
+// --- External imports ---
+import { artifacts, ethers, waffle } from 'hardhat';
+import { BigNumber, BigNumberish } from 'ethers';
+import { Artifact } from 'hardhat/types';
+import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/dist/src/signer-with-address';
+import { expect } from 'chai';
+
+// --- Our imports ---
+import { GrantRegistry } from '../typechain/GrantRegistry';
+
+// --- Parse and define helpers ---
+const { deployContract } = waffle;
+const { isAddress } = ethers.utils;
+const randomAddress = () => ethers.Wallet.createRandom().address;
+const randomPtr = () => String(Math.random()); // doesn't need to be a valid URL for testing
+const BN = (x: BigNumberish) => BigNumber.from(x);
+
+// --- Type definitions ---
+// The output from ethers/typechain allows array or object access to grant data, so we must define types for
+// handling the Grant struct as done below
+type GrantObject = {
+  id: BigNumber;
+  owner: string;
+  payee: string;
+  metaPtr: string;
+};
+type GrantArray = [BigNumber, string, string, string];
+type GrantEthers = GrantArray & GrantObject;
+type Grant = GrantObject | GrantEthers;
+
+// --- GrantRegistry tests ---
+describe('GrantRegistry', function () {
+  let deployer: SignerWithAddress, user1: SignerWithAddress, user2: SignerWithAddress;
+  let registry: GrantRegistry;
+
+  // Helper method to create a grant and return the data used to create it
+  async function createGrant(owner: string, payee: string, metaPtr: string, user?: SignerWithAddress) {
+    user = user ? user : deployer; // if a signer, `user`, is not specified we use the `deployer` account
+    await registry.connect(user).createGrant(owner, payee, metaPtr);
+    return { owner, payee, metaPtr };
+  }
+
+  // Helper method to verify that two Grant objects are equal
+  function expectEqualGrants(grant1: Grant, grant2: Grant) {
+    expect(grant1.id).to.equal(grant2.id);
+    expect(grant1.owner).to.equal(grant2.owner);
+    expect(grant1.payee).to.equal(grant2.payee);
+    expect(grant1.metaPtr).to.equal(grant2.metaPtr);
+  }
+
+  beforeEach(async () => {
+    [deployer, user1, user2] = await ethers.getSigners();
+
+    const grantRegistryArtifact: Artifact = await artifacts.readArtifact('GrantRegistry');
+    registry = <GrantRegistry>await deployContract(deployer, grantRegistryArtifact);
```

Comment:
this syntax looks a little strange to me. <...> is usually used to supply a type for a generic. so what is the generic here? Or is this some other language feature entirely?

Reply:
It's a way of type casting, so `<GrantRegistry>await deployContract()` is identical to `(await deployContract()) as GrantRegistry`

https://www.typescriptlang.org/docs/handbook/basic-types.html#type-assertions

Reply:
I was confused by this at first too, but I'm fine with using this syntax. Any objections @wildmolasses.

Reply:
is this an area where we value consistency? if so, for the following 3 reasons i recommend we use the "as" assertions over angle bracket assertions:

- we use the `as` syntax in all other repos we work on
- angle bracket syntax is not compatible with jsx
- angle brackets are already used for generics